File: client/internal/domain/evidence.go
package domain

type SecureHardwarePlatformReport interface {
	ReportFrom() SecureHardwarePlatform
}

type Evidence[R SecureHardwarePlatformReport] struct {
	HardwareEvidence HardwareEvidence[R]
	SoftwareEvidence SoftwareEvidence
}

type HardwareEvidence[R SecureHardwarePlatformReport] interface {
	SecureHardwarePlatform() SecureHardwarePlatform

	Report() R

	Raw() SignedRaw
}

// ! assuming TPM-based software evidence, therefore, no generics needed
type SoftwareEvidence interface {
	CloudServiceProvider() CloudServiceProvider

	Report() *TpmReport

	Raw() SignedRaw
}

File: server/src/domain/evidence.rs
pub struct EvidenceBundle {
    pub hardware: HardwareEvidence,
    pub software: SoftwareEvidence,
}

#[derive(Clone)]
pub struct HardwareEvidence {
    pub raw: Vec<u8>,
    pub cert: Vec<u8>,
}

#[derive(Clone)]
pub struct SoftwareEvidence {
    pub signed_raw: Vec<u8>,
    pub signature: Vec<u8>,
    pub cert: Vec<u8>,
}
